Support transactions in models

alflennik opened the pull request:
Apologies for the ginormous PR, believe it or not this is only half of the larger effort to implement transaction support.

The file change count is high but most of the changed files only have one line in their imports changed. Also the lines changed is inflated since there's a lot of duplicated code.

Here is a summary of the changes:

1. Duplicates the server/models/services folder into server/models/services.deprecated
2. Updates all imports of the model services to this new deprecated folder.

These changes allow the entire system to keep working without changes. This is the strategy I've used to break this larger effort into two PRs.

3. Updates the server/models/services with a new object-based API, making it way easier than the old positional arguments to supply transactions.
4. Updates the tests to support the new API. Where tests were missing, I added new tests.
5. Makes all the models files consistent, CollectionJobService and TestPlanService were both diverging from the other model files and that made this kind of change difficult.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206550059227038

howard-e commented on server/models/services/TestPlanReportService.js:
```diff
@@ -109,156 +112,167 @@ const testPlanAssociation = testPlanAttributes => ({
-    search,
-    filter = {},
+const getTestPlanReports = async ({
+    // search, // Nothing to search
```
Agreed

howard-e reviewed (changes requested):
This is great! Even though you've summarized the PR down to "duplicating `services` and updating imports", I think a bit more has gone into it than just that! I took note of all the JSDoc updates, restructuring files in some instances to better follow the structure of the `<model>Service.js` files and a bit of intentionality and care with doing away with older parameter structures.

Excited for when we can get rid of `services.deprecated`. We've needed this approach for a while!

My inline comments only reflect minor cleanup needs.

Paul-Clue reviewed (commented):
Hey Thanks for this work!
My main focus of this review is making sure that there are no missing or failing tests for the new object oriented structure of the API. As such, I found a few functions that may be missing tests:

`AtService.js`
    1. `removeAtVersionById`
    2. `getUniqueAtVersionsForReport`

`BrowserService.js`
    1. `removeBrowserVersionByQuery`
    2. `findOrCreateBrowserVersion`

`CollectionJobService.js`
    1. `scheduleCollectionJob` automation/graphql tests only
    2. `restartCollectionJob` automation/graphql tests only
    3. `cancelCollectionJob` automation/graphql tests only
    4. `retryCanceledCollections` graphql tests only

`TestPlanReportService.js`
    1. `updateTestPlanReportById`
    2. `removeTestPlanReportById`

`UserService.js`
    1. `getUserRoles`
    2. `getUserAts`
    3. `bulkGetOrReplaceUserAts`
    4. `deleteUserFromRole`

`TestsService.js`
    1. `getTests`

`TestResultWriteService.js`
    1. `findOrCreateTestResult` graphql / test plan run tests only

`TestResultReadService.js`
    1. `getTestResults`
    2. `getFinalizedTestResults`

`TestPlanService.js`
    1. `removeTestPlanById`

`TestPlanRunService.js`
    1. `removeTestPlanRunResultsByQuery`
    2. `getTestResultsUsingAtVersion`

Also the parameter (`t`) that you are using in the new API is not very descriptive. It seems like the `t` represents `transaction`. Maybe we can make this param more descriptive in a future PR.

alflennik commented:
Hi reviewers, changes since the last review are here: https://github.com/w3c/aria-at-app/pull/922/files/a904404adc1ed8da85d47fe604f233898dc0deb6..73b597c8400f1f3d19f28f49056a57914563badd

@Paul-Clue, thanks for listing out those missing tests, while implementing them I actually found a bug and a couple of mislabeled fields so that was a great note.

I didn't implement tests which were already covered elsewhere, but I made sure that all the methods which are universally tested in other services were tested.

About the `t` variable name, I had a whole speech prepared about how the name `t` is more efficient when transactions are so ubiquitous, but actually, thinking about it over a few days since your review, I've actually come around and I think you're right. So I renamed `t` to `transaction`.

alflennik commented on server/models/services/CollectionJobService.js:
```diff
@@ -1,22 +1,20 @@
 const ModelService = require('./ModelService');
+const { CollectionJob, sequelize } = require('../');
```
Woot!

alflennik commented on server/models/services/AtService.js:
```diff
@@ -52,493 +52,501 @@ const browserAssociation = browserAttributes => ({
- * @param {*} options.transaction - Sequelize transaction
+ * @param {object} options
+ * @param {object} options.where - values of the AtVersion record to be updated
+ * @param {object} options.values - values to be used to update columns for the record being referenced for {@param queryParams}
```
Done

alflennik commented on server/controllers/FakeUserController.js:
```diff
@@ -0,0 +1,36 @@
+    res.status(200).send('');
+};
+
+module.exports = setFakeUserController;
```
You're right, good catch. Done.

Paul-Clue commented on server/models/services/CollectionJobService.js:
```diff
@@ -518,24 +522,28 @@ const scheduleCollectionJob = async (
+    transaction
 }) => {
-    const existingJob = await getCollectionJobById({ id, t });
+    const existingJob = await getCollectionJobById({ id, transaction });
```
xxxx

Paul-Clue commented on server/models/services/CollectionJobService.js:
```diff
@@ -518,24 +522,28 @@ const scheduleCollectionJob = async (
 
 /**
- * Gets one CollectionJob and optionally updates it, or creates it if it doesn't exist.
+ * Gets one CollectionJob and optionally updates it, or creates it if it doesn'transaction exist.
```
A typo was created here.

alflennik replied:
Done, nice catch finding this in the voluminous changes 😅

Paul-Clue commented on server/models/services/TestPlanReportService.js:
```diff
@@ -309,45 +302,54 @@ const updateTestPlanReportById = async ({
 
 /**
- * Gets one TestPlanReport, or creates it if it doesn't exist, and then optionally updates it. Supports nested / associated values.
+ * Gets one TestPlanReport, or creates it if it doesn'transaction exist, and then optionally updates it. Supports nested / associated values.
```
Another typo

Paul-Clue commented on server/models/services/TestPlanRunService.js:
```diff
@@ -246,9 +246,9 @@ const createTestPlanRun = async ({
+    transaction
 }) => {
-    // shouldn't have duplicate entries for a tester unless it's automated
+    // shouldn'transaction have duplicate entries for a tester unless it's automated
```
Typo

Paul-Clue reviewed (approved):
It's good to go after fixing those tiny typos!
